Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,525 @@
+-                    boolean flag3 = a(chunkcoordintpair, i1, j1) <= this.viewDistance;
+-                    boolean flag4 = a(chunkcoordintpair, i, j) <= this.viewDistance;
++                    boolean flag3 = a(chunkcoordintpair, i1, j1) <= playerViewDistance;
++                    boolean flag4 = a(chunkcoordintpair, i, j) <= playerViewDistance;
+ 
+                     this.sendChunk(entityplayer, chunkcoordintpair, new Packet[2], flag3, flag4);
+                 }
+@@ -1340,8 +1350,8 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+             boolean flag5;
+             boolean flag6;
+ 
+-            for (k1 = i1 - this.viewDistance; k1 <= i1 + this.viewDistance; ++k1) {
+-                for (l1 = j1 - this.viewDistance; l1 <= j1 + this.viewDistance; ++l1) {
++            for (k1 = i1 - playerViewDistance; k1 <= i1 + playerViewDistance; ++k1) {
++                for (l1 = j1 - playerViewDistance; l1 <= j1 + playerViewDistance; ++l1) {
+                     chunkcoordintpair1 = new ChunkCoordIntPair(k1, l1);
+                     flag5 = true;
+                     flag6 = false;
+@@ -1349,8 +1359,9 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+                 }
+             }
+ 
+-            for (k1 = i - this.viewDistance; k1 <= i + this.viewDistance; ++k1) {
+-                for (l1 = j - this.viewDistance; l1 <= j + this.viewDistance; ++l1) {
++            for (k1 = i - playerViewDistance; k1 <= i + playerViewDistance; ++k1) {
++                for (l1 = j - playerViewDistance; l1 <= j + playerViewDistance; ++l1) {
++                    //Paper end
+                     chunkcoordintpair1 = new ChunkCoordIntPair(k1, l1);
+                     flag5 = false;
+                     flag6 = true;
+@@ -1366,7 +1377,7 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+         return this.playerMap.a(chunkcoordintpair.pair()).filter((entityplayer) -> {
+             int i = b(chunkcoordintpair, entityplayer, true);
+ 
+-            return i > this.viewDistance ? false : !flag || i == this.viewDistance;
++            return i > entityplayer.getTrackViewDistance() ? false : !flag || i == entityplayer.getTrackViewDistance(); //Paper ticket level pp-view-dist
+         });
+     }
+ 
+@@ -1616,7 +1627,7 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+             org.spigotmc.AsyncCatcher.catchOp("player tracker update"); // Spigot
+             if (entityplayer != this.tracker) {
+                 Vec3D vec3d = (new Vec3D(entityplayer.locX, entityplayer.locY, entityplayer.locZ)).d(this.tracker.getPositionVector()); // MC-155077, SPIGOT-5113
+-                int i = Math.min(this.trackingDistance, (PlayerChunkMap.this.viewDistance - 1) * 16);
++                int i = Math.min(this.trackingDistance, (entityplayer.getTrackViewDistance() - 1) * 16); //Paper ticket level pp-view-dist
```

Comment:
The problem with this change is we would send entities to players in this tracking range but not their corresponding chunks. Chunks only are sent to players if they are at ticket level 32 & have neighbours in a 1 chunk radius loaded. This is why it's viewDistance - 1, as the outside frame of a view distance square does not have their neighbours loaded

Reply:
noted, we can change this to the normal view distance

Reply:
changed and will be committed